### PR TITLE
Use cowboy to test multipart/form-data

### PR DESCRIPTION
**Pull request by jlouis**
This PR introduces a small test case for the multipart upload. Currently it fails with a 500 error inside cowboy because hackney is doing things which is against the RFC. The PR is part of fixing #197 since it provides the counter-example for the current operation.

I envision we can take this PR and a fix together so we can avoid regressions on this area later on.

### CODE REVIEW

**Comment by jlouis** [comment]
Analysis of the failing test case: the test fails because it uses rebar 2.5.x and it doesn't understand test profiles. Hence it doesn't understand the necessity of downloading cowboy to the system in order to make it work. Hopefully I can solve this by pushing an explicit test_deps section in rebar.config and make this go through locally. 

I think the other failing test case is a fallout from the teardown not being run correctly.

**Comment by jlouis** [comment]
Progress: Working down to R16B1 now. Current culprit is that ensure_all_started is not present there. It should be a fairly easy fix if you still want to track older releases.

**Comment by jlouis** [comment]
Still being trolled by R16B1 and earlier :)

**Comment by jlouis** [comment]
Down to R15B3 since `cowboy` has no support that far back. @benoitc should I disable this test for these old platforms and skip it?

**Comment by benoitc** [comment]
thanks! i am fine to disable the support of r16. i don't use it myself and next will probably be 17/18 only

**Comment by jlouis** [comment]
I've disabled R15 support and this works! R16 is still in there. Next up, I'll clean up the commits by rebasing, then we can merge this.

File: rebar.config.script [hunk 1]
```diff
@@ -21,7 +21,8 @@ Rebar2Deps = [{idna, ".*",
 
               {ssl_verify_hostname, ".*",
                {git, "https://github.com/deadtrickster/ssl_verify_hostname.erl",
-                {tag, "1.0.5"}}}],
+                {tag, "1.0.5"}}},
+              {cowboy, ".*", {git, "https://github.com/ninenines/cowboy.git", {tag, "1.0.4"}}}],
```
Comment (benoitc) [inline rebar.config.script]: maybe we should test the env here for example if a TEST has been set in the env. So we don't need to fetch cowboy all the time? Thoughts?

**Comment by jlouis** [comment]
Rebased this to be a better and cleaner piece of work.

I agree we could test for the env if TEST has been set. It cut avoid the cowboy dependency by default. What is the idea in rebar 2.x? Is it already set, or do we have to supply it in some way when the script runs?

**Comment by benoitc** [comment]
hrm I don't think anything exists in rebar2 . Imo we could juste set the
env and pass it by default. I can do the change anyway :)

**Comment by jlouis** [comment]
Nice, then merge this. It is ready and partially solves #197 It doesn't do anything about the content disposition part of that issue.

**Comment by edgurgel** [comment]
I wonder if we could use the existing `httpbin` instance that runs on travis for the integration tests (sadly they do not run on rebar3 https://github.com/benoitc/hackney/pull/202) instead of a new cowboy server.

**Comment by benoitc** [comment]
@edgurgel good idea :)

**Comment by jlouis** [comment]
You *can* run rebar3 if you put it into the repository itself and call it. The problem with this approach is the obvious: compiled a newer version of rebar3? You can't run it on the older releases and vice versa.

The other problem we face is that of including cowboy or not. The advantage of including it is you can run tests locally, since the `httpbin` instance might not be there locally... One has to make this work in order to make local development possible.

I've also seen integration tests are not running on `rebar3`. This is because there is no module named `hackney_integration` and hence the tests in `hackney_integration_tests` have no backing module and should not be run. Rebar 2.x does something "magical" here which frankly may have confused people more than it has helped.

**Comment by benoitc** [comment]
So adding rebar3 inside travis can be done using such script:
https://github.com/barrel-db/hooks/blob/master/support/bootstrap_travis.sh

The script is launch before installing:
https://github.com/barrel-db/hooks/blob/master/.travis.yml#L5-L6

For HTTPBIN I guess it could also be automated but I see the point. Maybe we could replace httpbin then but it will require some work. Sometimes ago I wrote [echohttp:](https://github.com/benoitc/echohttp) which can be a good start.

I think it is a different issue anyway :) I propose to merge that PR, do the change in rebar config (checking if the environnement setting `TEST` is existing) and create a new issue to discuss further about it. Thoughts?

**Comment by jlouis** [comment]
Yes, this change should go in, then get the ball rolling forward with fixes as they occur.

**Comment by benoitc** [comment]
merged. Thanks!

**Pull request merged**